### PR TITLE
sql: remove testing cluster settings in favor of metamorphic build tag

### LINKED ISSUE #51156: sql:  unexpected error from the vectorized engine: index out of range [2] with length 2
I've hit the following error on [one of my builds](https://teamcity.cockroachdb.com/viewLog.html?buildId=2074552&buildTypeId=Cockroach_UnitTests)
I doubt that I'm causing it. Didn't find any other open issues with the error. Doesn't repro immediately.

```
=== RUN   TestLogic/fakedist-disk/tuple/labeled_tuple_column_access_errors
    TestLogic/fakedist-disk/join: logic.go:2298: 
        
        testdata/logic_test/join:296: SELECT o.x, t.y FROM onecolumn o INNER JOIN twocolumn t ON (o.x=t.x AND t.y=53)
        expected success, but found
        (XX000) internal error: unexpected error from the vectorized engine: runtime error: index out of range [2] with length 2
        error.go:91: in func1()
        DETAIL: stack trace:
        github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:91: func1()
        runtime/panic.go:969: gopanic()
        runtime/panic.go:88: goPanicIndex()
        github.com/cockroachdb/cockroach/pkg/sql/colencoding/value_encoding.go:95: decodeUntaggedDatumToCol()
        github.com/cockroachdb/cockroach/pkg/sql/colencoding/value_encoding.go:46: DecodeTableValueToCol()
        github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:1182: processValueBytes()
        github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:1207: processValueTuple()
        github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:952: processValue()
        github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:741: nextBatch()
        github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:588: nextAdapter()
        github.com/cockroachdb/cockroach/pkg/sql/colmem/allocator.go:235: PerformOperation()
        github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:580: NextBatch()
        github.com/cockroachdb/cockroach/pkg/sql/colfetcher/colbatch_scan.go:71: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/cancel_checker.go:48: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/simple_project.go:124: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:42: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/disk_spiller.go:326: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/external_hash_joiner.go:495: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/disk_spiller.go:212: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/simple_project.go:124: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:42: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/deselector.go:55: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:245: func1()
        github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:94: CatchVectorizedRuntimeError()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:262: sendBatches()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:338: runWithStream()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:192: Run()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:559: func1()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1121: 1()
        runtime/asm_amd64.s:1373: goexit()
```

cc @yuzefovich

### LINKED ISSUE #57088: colexec: panic on index out of bounds
Seen in https://teamcity.cockroachdb.com/viewLog.html?buildId=2471003&tab=buildResultsDiv&buildTypeId=Cockroach_UnitTests_Test

```
=== RUN   TestLogic/fakedist-disk/aggregate/bit_aggregates
--- progress: testdata/logic_test/sequences: 163 statements/queries
=== CONT  TestLogic/fakedist-disk/aggregate/bit_aggregates
    logic.go:2419: 
        testdata/logic_test/aggregate:3158: SELECT bit_or(v), bit_or(b) FROM vals
        expected success, but found
        (XX000) internal error: unexpected error from the vectorized engine: runtime error: index out of range [1] with length 1
        error.go:90: in func1()
        DETAIL: stack trace:
        github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:90: func1()
        runtime/panic.go:969: gopanic()
        runtime/panic.go:88: goPanicIndex()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecagg/ordered_default_agg.eg.go:59: func1()
        github.com/cockroachdb/cockroach/pkg/sql/colmem/allocator.go:292: PerformOperation()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecagg/ordered_default_agg.eg.go:56: Compute()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/aggregators_util.go:117: performAggregation()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/ordered_aggregator.go:272: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:44: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/deselector.go:61: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:240: func1()
        github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:93: CatchVectorizedRuntimeError()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:232: sendBatches()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:316: runWithStream()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:180: Run()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:679: func1()
        github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1281: 1()
        runtime/asm_amd64.s:1374: goexit()
        NOTE: internal errors may have more details in logs. Use -show-logs.
    logic.go:2164: 
         pq: internal error: unexpected error from the vectorized engine: runtime error: index out of range [1] with length 1
    logic.go:1720: 
         testdata/logic_test/aggregate:3163: too many errors encountered, skipping the rest of the input
--- done: testdata/logic_test/aggregate with config fakedist-disk: 505 tests, 3 failures
--- done: testdata/logic_test/system with config fakedist-disk: 58 tests, 0 failures
            --- FAIL: TestLogic/fakedist-disk/aggregate/bit_aggregates (0.18s)
```

### LINKED ISSUE #57198: colexec: index out of range panic from vectorized engine
Saw the following index out of range [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=2478639&buildTypeId=Cockroach_UnitTests).

```
=== CONT  TestLogic/fakedist-spec-planning/inverted_join_json_array
    logic.go:2419: 
        
        testdata/logic_test/inverted_join_json_array:320: SELECT * FROM
        (SELECT j1.a, j2.a FROM json_tab@foo_inv AS j1, json_tab AS j2 WHERE j1.b @> j2.b) AS inv_join(a1, a2)
        FULL OUTER JOIN
        (SELECT j1.a, j2.a FROM json_tab@primary AS j1, json_tab AS j2 WHERE j1.b @> j2.b) AS cross_join(a1, a2)
        ON inv_join.a1 = cross_join.a1 AND inv_join.a2 = cross_join.a2
        WHERE inv_join.a1 IS NULL OR cross_join.a1 IS NULL
        expected success, but found
        (XX000) internal error: unexpected error from the vectorized engine: runtime error: index out of range [3] with length 3
        error.go:90: in func1()
        DETAIL: stack trace:
        github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:90: func1()
        runtime/panic.go:969: gopanic()
        runtime/panic.go:88: goPanicIndex()
        github.com/cockroachdb/cockroach/pkg/col/coldata/vec.eg.go:620: Copy()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:570: func1()
        github.com/cockroachdb/cockroach/pkg/sql/colmem/allocator.go:292: PerformOperation()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:564: congregate()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:546: exec()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/hashjoiner.go:281: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/disk_spiller.go:201: func1()
        github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:93: CatchVectorizedRuntimeError()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/disk_spiller.go:199: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/operator.go:408: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/buffer.go:62: advance()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/case.go:110: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/bool_vec_to_sel.go:150: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/bool_vec_to_sel.go:57: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/simple_project.go:124: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:44: Next()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/parallel_unordered_synchronizer.go:195: 1()
        github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:93: CatchVectorizedRuntimeError()
        github.com/cockroachdb/cockroach/pkg/sql/colexec/parallel_unordered_synchronizer.go:229: func2()
        runtime/asm_amd64.s:1374: goexit()
        
        NOTE: internal errors may have more details in logs. Use -show-logs.
    logic.go:2164: 
         pq: internal error: unexpected error from the vectorized engine: runtime error: index out of range [3] with length 3
--- done: testdata/logic_test/inverted_join_json_array with config fakedist-spec-planning: 8 tests, 2 failures
    logic.go:2927: 
        testdata/logic_test/inverted_join_json_array:329: error while processing
    logic.go:2927: testdata/logic_test/inverted_join_json_array:329: too many errors encountered, skipping the rest of the input
        --- FAIL: TestLogic/fakedist-spec-planning/inverted_join_json_array (1.18s)
```

This was on #57155, rebased atop e9d66e5dee429f15d5d27e0f4fa6294f8b9897e4 at the time. I think my PR is unrelated to the panic above.

### PR DESCRIPTION
**Pull request by yuzefovich**
**coldata: fix tests for randomized batch size**

This commit makes all tests to work when the batch size is randomized.
Additionally, it removes `nulls.swap` function that is not used (we can
always look up the implementation in the history if the need arises).

Release note: None

**util: fix metamorphic helper in an edge case**

Previously, if the specified range contained exactly one integer (i.e.
`min == max`), we would hit a division by zero error. This commit fixes
the issue.

Release note: None

**sql: fix all failures when mutations.MaxBatchSize is randomized**

This commit fixes all the test failures that occur when running with
metamorphic tag and setting `mutations.MaxBatchSize` to 1 explicitly -
a couple of unit tests need to be skipped as well as some of the logic
tests.

The logic tests now support `!metamorphic` blocklist directive that
skips the whole file under the metamorphic build. This is only necessary
for very few cases (mostly around tracing of the KV operations).

Release note: None

**sql: remove testing.vectorize.batch_size setting**

With the recent addition of the metamorphic build tag we no longer need
to expose the `sql.testing.vectorize.batch_size` cluster setting in
order to randomize `coldata.BatchSize()` for testing purposes, so this
commit retires the setting. The only testing coverage we're losing is
`tpchvec/smallbatchsize` test, but I don't think it ever found any
issues, so it doesn't seem important to figure out how to use the
metamorphic tag for it.

Fixes: #51156.
Fixes: #57088.
Fixes: #57198.
Addresses: #55348.

Release note: None (removing non-public setting)

**sql: remove testing.mutations.max_batch_size setting**

This commit removes no longer used
`sql.testing.mutations.max_batch_size` cluster setting since the
metamorphic build provides us the coverage the setting was intended to
(it was currently skipped anyway in all cases).

Additionally it moves a single logic test file from execbuilder tests
(to undo the move when we introduced the cluster setting).

Addresses: #55348.

Release note: None (removing non-public setting)

**util,build: use `crdb_test` tag instead of `metamorphic`**

This commit switches the builds to become "metamorphic" based on the
recently introduced `crdb_test` tag (which is appended to all test
targets) instead of `metamorphic`. The corresponding teamcity config is
removed because we will get the metamorphic runs with the regular test
runs now.

Note that since some logic tests cannot be run when some constants
initialized to non-default value, in order to not lose the test coverage
for those tests, even if `crdb_test` build is specified, with 20%
probability the build will not be metamorphic.

Additionally, this commit switches the logic of
`ConstantWithMetamorphic*` methods to also choose the default value (in
25% cases). The reasoning for this change is that we want to make sure
to run the tests with the default values relatively often.

To summarize, the current behavior is:
- `crdb_test` build tag is specified (all test targets):
  - with 20% probability the build is not metamorphic, so all magic
constants get the default value.
  - with 80% probability the build is metamorphic. Every value being
metamorphized has 75% of being initialized to non-default value.
- `crdb_test` build is not specified:
  - the build is not metamorphic, so all magic constants get the
produciton value.

Another minor change is a fix to
`row.TestingSetDatumRowConverterBatchSize` to properly reset a value to
its old one in the returned closure.

`TestMysqldumpDataReader` is currently skipped when the metamorphic
build occurs pending further investigation.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
Ok, I think I fixed all of the current failures of the metamorphic CI config. I also confirmed that using value 1 for `mutations.MaxBatchSize` would have caught the upsert bug from a couple of months ago.

I updated the metamorphic config to run nightly on master, the only remaining thing is making sure that github issues are filed.

**Review by asubiotto — CHANGES_REQUESTED** [review]
Great to see this, thanks!

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 13 of 13 files at r3, 8 of 8 files at r4, 9 of 9 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-:-MO0gN1Z-WW10qR88WZC:b-9os38)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/logictest/logic.go, line 1622 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-MO0ht4137G5mZJK8mi2:-MO0ht424rbW1QSzc_oR:bm0i9dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac624534e59391ad409dcac1f1b63e48212b6c0a/pkg/sql/logictest/logic.go#L1622)):*
> ```Go
> 		for c := range blocklist {
> 			if c == "metamorphic" {
> 				return configs
> ```

I don't understand this logic, could you explain it to me? Basically if there is a `!metamorphic` in the file we'll run with no `configs`?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/logictest/logic.go, line 1622 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-MO0ht4137G5mZJK8mi2:-MO1xcY2-YpZKYUem7Gp:b81pzc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac624534e59391ad409dcac1f1b63e48212b6c0a/pkg/sql/logictest/logic.go#L1622)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't understand this logic, could you explain it to me? Basically if there is a `!metamorphic` in the file we'll run with no `configs`?
</blockquote></details>

Yeah, exactly, added a comment.

**Review by asubiotto — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/logictest/logic.go, line 1622 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-MO0ht4137G5mZJK8mi2:-MO2AmrLEUPEkCwrpIJB:bdv38gk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac624534e59391ad409dcac1f1b63e48212b6c0a/pkg/sql/logictest/logic.go#L1622)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yeah, exactly, added a comment.
</blockquote></details>

But is that right? Don't we want to run with the configs that weren't blocklist directives or the default set if all the configs were blocklist directives?

e.g.
```
# LogicTest: local !metamorphic
```
Should still run with `local`
```
# LogicTest: !3node-tenant !metamorphic
```
Should run with all default configs minus `3node-tenant` and `metamorphic`

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/logictest/logic.go, line 1622 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-MO0ht4137G5mZJK8mi2:-MO2OUkI2xvzYuHoUCE0:b47bxod) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac624534e59391ad409dcac1f1b63e48212b6c0a/pkg/sql/logictest/logic.go#L1622)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

But is that right? Don't we want to run with the configs that weren't blocklist directives or the default set if all the configs were blocklist directives?

e.g.
```
# LogicTest: local !metamorphic
```
Should still run with `local`
```
# LogicTest: !3node-tenant !metamorphic
```
Should run with all default configs minus `3node-tenant` and `metamorphic`
</blockquote></details>

Extended the comment, PTAL.

**Comment by yuzefovich** [comment]
We had a report of the metamorphic build failure [here](https://github.com/cockroachdb/cockroach/issues/54753#issuecomment-741378460) which I triggered manually. I removed the old metamorphic build config and created a new one under `Nightlies` project which will run nightly. I think the only remaining question is how to filter the branches on which it'll run (I think as is it'll run on all release branches, including master), but running on branches older than the master is useless since they don't support the metamorphic tag.

**Comment by RaduBerinde** [comment]
Why not enable the metamorphic stuff on all test targets? (the `crdb_test` tag I just added in #57532).

**Comment by yuzefovich** [comment]
Hm, I think we want to run the tests both with and without the randomizations, and currently `crdb_test` tag is appended to all test configs, so we would only run with the randomizations if we were to treat it as "metamorphic=true".

I guess we could adjust the randomizer so that in 50% cases it used the default value, and in other 50% it would use the randomized value. That sounds good to me. @asubiotto @jordanlewis thoughts?

**Comment by mgartner** [comment]
If you make it random, then any test that fails with metamorphic configs will be a flaky failure rather than deterministic. Will that cause confusion?

**Comment by yuzefovich** [comment]
The chosen random values are logged at the test startup, so ideally all failures would be reproducible (at least under stress). I guess we'll have to see.

I think overall we do have many magic constants throughout the code base, and it'd be very beneficial to randomize them during some test runs. It's possible that multiple randomizations combined could trigger something unexpected, but I hope the signal won't be that noisy.

**Comment by RaduBerinde** [comment]
IMO it's preferable for as much randomization as reasonably possible to happen during regular test runs. If it only happens in a nightly run, whoever sets up that nightly run is stuck monitoring it and debugging/triaging failures.

**Comment by yuzefovich** [comment]
That's a good point. I've been debating in my mind whether the metamorphic tests should be run as part of CI or not.

A minor complication here is that some tests rely on a particular default value of the constants for its expected output, so we'll need to be able to tell whether the randomizer chose the default or random value.

**Comment by yuzefovich** [comment]
Added another commit to use `crdb_test` tag. PTAL.

**Comment by yuzefovich** [comment]
Yay! The build is green, and I can see in the logs that the constants have been randomized.

**Review by asubiotto — COMMENTED** [review]
Does it make sense to keep the `metamorphic` blocklist directive in that case? Would it be better to rename it `!crdb_test` because that's truly what it does?

> Reviewed 25 of 25 files at r6, 1 of 1 files at r7, 13 of 13 files at r8, 8 of 8 files at r9, 9 of 9 files at r10, 4 of 4 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-:-MOBJPEj7EgnaXiutgwU:bord3b2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

**Review by RaduBerinde — APPROVED** [review]
It would be pretty strange to have logic test files that never run as part of the logic tests, unless you run them with `go test` manually, no?

It would be ideal if the test could disable specific metamorphic changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-:-MOCd7-i-_UFT5GBXvSW:b-c3dwjc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

**Comment by yuzefovich** [comment]
Yeah, with Alfonso's comment I realized that we need to adjust the logic test directive. I think I'll change the randomization approach so that with some probability (let's say 10%) only the default values are chosen for all magic constants, and then we would set a global flag indicating that. The blocklist directive would only be applied if that flag is not set.

**Review by RaduBerinde — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/util/metamorphic_off.go, line 20 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-MOCeAfSD2OTL3Hoze_m:-MOCeAfSD2OTL3Hoze_n:bvuiwkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2ba6d29f19144b4755b0563eb85f3233230acea/pkg/util/metamorphic_off.go#L20)):*
> ```Go
> // metamorphic-style perturbations that will not affect test results but will
> // exercise different parts of the code.
> const MetamorphicBuild = false
> ```

[nit] I would change this to TestBuild or CrdbTestBuild (and the file names to crdb_test_on/off), because it will be useful for other things. We can have a separate `const MetamorphicBuild = CrdbTestBuild` in `constants.go`

**Review by RaduBerinde — APPROVED** [review]
Not sure I understand - I assumed that the tests that need `!metamorphic` would fail with non-default values?

Btw I noticed that `TestingSetDatumRowConverterBatchSize` restores the value to the default, not the previous (potentially randomized) value.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-:-MOCh4JZ4s107c9of7Pn:bdq1cwp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

**Comment by yuzefovich** [comment]
Yes, the logic tests with `!metamorphic` directive will fail with non-default magic constants (in particular `mutations.MaxBatchSize`), so all such logic test files are skipped if the build is determined to be "metamorphic".

I think we need to adjust the skipping logic to apply only if `crdb_test` is specified *and* non-default value is used at least for one constant.

**Comment by RaduBerinde** [comment]
It would be ideal if we could reset the constants from the logic tests and always run the test. Otherwise I can put garbage in a test file, run the tests, and some of the time it would look like everything is ok. If I'm trying to debug something I will need to keep running multiple times each time; having each engineer discover this subtlety will add up to a lot of wasted time.

I realize it may be a tall order.. we could use things like `TestingSetDatumRowConverterBatchSize` but they would be problematic for parallel tests.

**Comment by RaduBerinde** [comment]
If it's really only one or two constants that are relevant here, we should plumb their values through server config, and perhaps even allowing changing it via a SQL statement (and the tests that care would do that rather than relying on a tag).

**Comment by RaduBerinde** [comment]
Note that I found `SET CLUSTER SETTING sql.testing.vectorize.batch_size = X` very useful to debug failures that were dependent on the batch size, so that's another argument for providing something like that for other constants.

**Comment by yuzefovich** [comment]
I think the only special constant is `mutations.MaxBatchSize` since it changes the boundaries of KV batches which changes the tracing.

We want to remove these cluster settings because we have found the cluster setting propagation to be the likely cause of very hard to debug test flakes (see the issues linked in the PR description).

**Comment by yuzefovich** [comment]
I updated the last commit.

The behavior is:
- `crdb_test` build tag is specified (all test targets):
  - with 20% probability the build is not metamorphic, so all magic
constants get the default value.
  - with 80% probability the build is metamorphic. Every value being
metamorphized has 75% of being initialized to non-default value.
- `crdb_test` build is not specified:
  - the build is not metamorphic, so all magic constants get the
produciton value.

**Comment by yuzefovich** [comment]
Once again I had to debug failures in this [build](https://teamcity.cockroachdb.com/viewLog.html?buildId=2508660&buildTypeId=Cockroach_UnitTests) the root cause of which - I believe - is the fact that we're randomizing `coldata.BatchSize()` via a cluster setting when running logic tests (stressing one logic tests on the branch I hit a failure within a minute when batch size is randomized to 1, but if I manually set the batch size to 1, then no failures in over 5 minutes).

**Comment by asubiotto** [comment]
> Once again I had to debug failures in this build the root cause of which - I believe - is the fact that we're randomizing coldata.BatchSize() via a cluster setting when running logic tests (stressing one logic tests on the branch I hit a failure within a minute when batch size is randomized to 1, but if I manually set the batch size to 1, then no failures in over 5 minutes).

I'm confused. Are we keeping the batch size cluster setting? I thought the point of this change was to remove these cluster settings (at least this is still called out in the commit messages). Might be easier/quicker to discuss this PR in our meeting today.

**Comment by yuzefovich** [comment]
Sorry, I should have been more clear - the build I linked above is from another PR (#55909), and my point of mentioning that failure was that I believe having the testing cluster settings (at least in logic tests) is very fragile and could produce very hard to debug flakes, and, therefore, I'm advocating for removal of the settings.

**Review by asubiotto — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57483#-:-MOa5OUdAZQqUwRxFQlY:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

**Comment by yuzefovich** [comment]
TFTR!

bors r+

**Comment by yuzefovich** [comment]
Will rebase on top of the tracing fix.

bors r-

**Comment by yuzefovich** [comment]
Hm, I've seen `importccl` package fail several times on this PR, and it failed again, so I'll need to take a closer look before merging it.

bors r-

**Comment by yuzefovich** [comment]
There was a valid timeout that I can easily reproduce, and it's possible that there is a bug with mysql dump import, I filed #57955 and skipped the test under metamorphic build for now.

bors r+

**Comment by RaduBerinde** [comment]
A particularly annoying side-effect is that when you run execbuilder tests with `-rewrite`, some of the files don't actually get rewritten. This is very annoying for optimizer work, where a very common workflow is to run -rewrite on all opt tests and study the diffs.

Is it possible to at least add a flag (or build tag?) that disables the metamorphic stuff? Or maybe the right solution, at least for the optimizer, is to move out all the tests that involve traces somewhere else.

**Comment by yuzefovich** [comment]
Yeah, good point. I don't think adding a flag will work (I'm assuming you're thinking about `TESTFLAGS` for logic tests) since they are processed too late, but we could add a build tag. What do you think about `crdb_test_off` flag that will need to be added when `-rewrite` option is used? The example invocation would be `make testoptlogic TESTFLAGS='-rewrite' TAGS=crdb_test_off`.
